File: 3514-shortest-distance-after-road-addition-queries-ii/shortest-distance-after-road-addition-queries-ii.java
class Solution {
    static class DSU {
        private int[] parent;
        private int[] size;
        private int numCC;

        public DSU(int n) {
            parent = new int[n];
            size = new int[n];
            numCC = n;
            for (int i = 0; i < n; i++) {
                parent[i] = i;
                size[i] = 1;
            }
        }

        public int find(int v) {
            if (v != parent[v]) {
                parent[v] = find(parent[v]);
            }
            return parent[v];
        }

        public void unite(int u, int v) {
            u = find(u);
            v = find(v);
            if (u != v) {
                numCC--;
                if (u < v) {
                    int temp = u;
                    u = v;
                    v = temp;
                }
                parent[v] = u;
                size[u] += size[v];
            }
        }

        public int getNumCC() {
            return numCC;
        }
    }

    public int[] shortestDistanceAfterQueries(int n, int[][] queries) {
        DSU dsu = new DSU(n);
        int[] ans = new int[queries.length];

        for (int i = 0; i < queries.length; i++) {
            int[] q = queries[i];
            int L = dsu.find(q[0]);
            int R = dsu.find(q[1] - 1);

            for (int j = L; j < R;) {
                if (dsu.find(j) + 1 <= R) {
                    dsu.unite(j, dsu.find(j) + 1);
                } else {
                    break;
                }
                j = dsu.find(j);
            }

            ans[i] = dsu.getNumCC() - 1;
        }

        return ans;
    }
}

File: 3514-shortest-distance-after-road-addition-queries-ii/shortest-distance-after-road-addition-queries-ii.cpp
class Solution {
public:
struct dsu {
    vector<int> parent, s;
    int numCC;

    dsu(int n) {
        parent = vector<int>(n);
        numCC = n;
        s = vector<int>(n, 1);
        for (int i = 1; i < n; ++i)
            parent[i] = i;
    }

    int find(int v) {
        if (v == parent[v])
            return v;
        return parent[v] = find(parent[v]);
    }

    void unite(int u, int v) {
        u = find(u);
        v = find(v);
        if (u != v) {
            numCC--;
            if (u < v)
                swap(u, v);
            parent[v] = u;
            s[u] += s[v];
        }
    }

    };

    vector<int> shortestDistanceAfterQueries(int n, vector<vector<int>>& queries) {
        dsu d(n);

        vector<int> ans;
        for (auto q: queries){
            int L = d.find(q[0]);
            int R = d.find(q[1]-1);
            
            for (int i = L; i < R;) {
                if (d.find(i) + 1 <= R)
                    d.unite(i, d.find(i) + 1);
                else
                    break;
                i = d.find(i);
            }

            ans.emplace_back(d.numCC-1);
        }

        return ans;
        }
};